# Probability, Sampling Distributions, and Random Variables

## Package Imports

We'll import three Python packages that we've used so far and will need for this assignment.  Those packages are pandas, matplotlib.pyplot, and seaborn.  

Run the cell provided below to access the functions in these packages.

You may also need to read in additional packages below.

In [1]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

<hr>

## 1. Champaign County 

**a)** First, read in the county.csv file.

In [2]:
county = pd.read_csv('county.csv')

Sample 1: We will start by focusing on our county (Champaign County in Illinois).  Because our county is home to UIUC, in which a lot of students and young adults live in apartments, we believe that our county will have a high percent of housing units in multi-unit structures.

**b)** Display just the row for Champaign County, Illinois.

Tip: We can use the **`&`** ("and") operator to indicate that we want **both** conditions on either side of the operator to be met.  We can use the **`|`** ("or") operator to indicate that we want **at least one** of the conditions to be met.  We can also chain these operators together if we need to represent more complex operations.

In [3]:
county[(county['name'] == 'Champaign County') & (county['state'] == 'Illinois')]

name     state pop2000  pop2010 pop2017 pop_change poverty  \
604  Champaign County  Illinois  179669   201081  209399       1.83    21.4   

     homeownership  multi_unit unemployment_rate metro    median_edu  \
604           55.9        35.2              4.24   yes  some_college   

    per_capita_income median_hh_income       smoking_ban  
604          29619.49            49586  data unavailable

**c)** Calculate the probability that a randomly selected county in the United States would have a multi-unit rate that is at least as high as Champaign county's rate.

In [4]:
# multi-unit rate of Champaign County's rate is 35.2
print("The probability that a randomly selected county in the United States would have a multi-unit rate that is at least as high as Champaign county's rate is" , county[county['multi_unit'] >= 35.2 ].shape[0] / county.shape[0])

The probability that a randomly selected county in the United States would have a multi-unit rate that is at least as high as Champaign county's rate is 0.02864417568427753


**d)** Calculate the probability that a randomly selected county in Illinois would have a multi-unit rate that is at least as high as Champaign county's rate.

In [5]:
# 'a' = county in Illinois that have a multi-unit rate that is at least as high as Champaign county's rate
a = county[(county['multi_unit'] >= 35.2) & (county['state'] == 'Illinois') ].shape[0]

# 'b' = number of counties in Illinois
b = county[county['state'] == 'Illinois'].shape[0]

In [6]:
print('The probability that a randomly selected county is Illinois would have a multi-unit rate that is at least as high as Champaign county is ', a/b)

The probability that a randomly selected county is Illinois would have a multi-unit rate that is at least as high as Champaign county is  0.0196078431372549


**e)** Are the events "has a multi-unit rate at least as high as Champaign County, Illinois" and "county is in the state of Illinois" independent for the counties in the United States?  Explain.

No, they are not independent events.If they were independent, the probability that a randomly selected county in the United States would have a multi-unit rate that is at least as high as Champaign's county's rate (part 1-c) should be as same as the probability that a randomly selected county in Illinois would have a multi-unit rate that is at least as high as Champaign's county's rate (part 1-d).

## 2. Describing Champaign + Neighboring Counties 

Sample 2: What if we extend from our county to include four neighboring counties (Vermilion, Ford, Piatt, and Douglas)?  will the average multi-unit rate for these five counties be higher than the average rate for five randomly selected counties from the population of U.S. counties?

**a)** To start, create a data frame that is composed of just the five following Illinois counties:

- Champaign
- Vermilion
- Ford
- Piatt
- Douglas

Print the resulting data frame.

In [7]:
Illinois_counties = county[county['state'] == 'Illinois']
neighboring_counties = Illinois_counties[ (county['name'] == 'Champaign County') |(county['name'] == 'Vermilion County') | (county['name'] == 'Ford County') | (county['name'] == 'Piatt County') | (county['name'] == 'Douglas County')]
neighboring_counties

C:\Users\happy\AppData\Local\Temp\ipykernel_35296\571762022.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  neighboring_counties = Illinois_counties[ (county['name'] == 'Champaign County') |(county['name'] == 'Vermilion County') | (county['name'] == 'Ford County') | (county['name'] == 'Piatt County') | (county['name'] == 'Douglas County')]


name     state pop2000  pop2010 pop2017 pop_change poverty  \
604  Champaign County  Illinois  179669   201081  209399       1.83    21.4   
615    Douglas County  Illinois   19922    19980   19748      -0.57    12.1   
621       Ford County  Illinois   14241    14081   13280      -2.86    15.7   
668      Piatt County  Illinois   16365    16729   16445       0.09     5.4   
686  Vermilion County  Illinois   83919    81625   77909      -3.31    19.8   

     homeownership  multi_unit unemployment_rate metro    median_edu  \
604           55.9        35.2              4.24   yes  some_college   
615           78.3        11.2              4.01    no    hs_diploma   
621           79.1         7.9               4.7   yes  some_college   
668           81.7         7.5              3.99   yes  some_college   
686           71.3        14.5              6.29   yes    hs_diploma   

    per_capita_income median_hh_income       smoking_ban  
604          29619.49            49586  data unavailable  
615          25796.91            52261           partial  
621          27234.78            50851              none  
668          33598.78            67360  data unavailable  
686          23759.02            44930  data unavailable

**b)** Calculate the mean multi-unit rate for these five counties.

In [8]:
neighboring_counties['multi_unit'].mean()

15.260000000000002

**c)** Calculate the mean and standard deviation for the multi-unit rate for the population of counties in the United States.

In [9]:
print(county['multi_unit'].mean())
print(county['multi_unit'].std())

12.321896880967534
9.290204854486706


**d)** Generate a random sample with replacement of size $n=5$ counties from the United States, and calculate the mean multi-unit rate for these 5 counties.  What symbol should you use to describe the value you just calculated?  Compare your calculated value to the values previously calculated in **2b** and **2c**.

In [10]:
county.sample(5, replace = True)['multi_unit'].mean()

13.959999999999999

We use X bar to describe the mean value of a sample. 
The mean multi-unit rate for this random sample of size n=5 is 19.74(part 2-d) and this is larger than the values previously calculated in 2b (15.26) and 2c (12.322). 

## 3. Comparing Champaign + Neighboring Counties 

**a)** The following MCmeans function has been provided for you.  Run the code in the first cell to load this function into Python.

Then, using the MCmeans function, geneerate output according to the following specifications:

- Specify the argument (input) for df to reference your data frame of counties
- Specify the argument for var to reference the variable name for multi-unit housing rates in the counties data frame
- Specify the argument to generate samples of size 5
- Specify the argument to generate 5000 repeated samples
- Save the output of this function to a Python object

In [11]:
def MCmeans(df, x='', replace=True, n=1, M=1):
    #INPUT:
    # df is a data frame
    # x is a text-valued name for a variable in the data frame
    # replace = True or False depending on whether 
    #    draws are with or without replacement
    # n = number of draws per sample
    # M = number of samples to draw
    MCstats = []
    for i in range(M):
        #1. Collect a random sample of size n with replacement
        #2. Take the mean of this random sample
        #3. Append this random sample mean to the SampleMeans list 
        # (which is our SAMPLING DISTRIBUTION OF SAMPLE MEANS!)
        MCstats.append(df[x].sample(n, replace=replace).mean())
    #4. returns the sampling distribution in a dataframe format
    return pd.DataFrame({x: MCstats})

In [12]:
sample_mean = MCmeans(county, x = 'multi_unit', replace = True, n = 5, M = 5000)
sample_mean

multi_unit
0          12.96
1           8.46
2          11.74
3          16.68
4          20.18
...          ...
4995       18.90
4996       13.16
4997        9.28
4998       31.70
4999       13.46

[5000 rows x 1 columns]

**b)** Calculate the mean and standard deviation for the object that is created in part **3a**.  Compare these values to those from **2c**.

In [13]:
print(sample_mean['multi_unit'].mean())

12.496751999999999


In [14]:
print(sample_mean['multi_unit'].std())

4.388076518658571


The result of part 2c is 12.321896880967534 as a mean and 9.290204854486706 as standard deviation. The result of part 3a is 12.4968 as a mean and 4.3880 as standard deviation. 
The mean value is quite similar to each other while a standard deviation for the object that is created in part 3a is smaller than std from part 2c. 

This is because the mean of population (mu) is same as the mean of sample (X bar). However, the standard deviation of sample (s) is a value of std of population (sigma) divided by square root n (sample size). Therefore, the std of part 3b must be approximately square root 5 times smaller than that of part 2c. 

**c)** What proportion of samples of size 5 have a mean multi-unit rate at least as large as the mean multi-unit rate of Champaign and four of its neighboring counties (from **2b**)?

In [16]:
# The mean multi-rate of Champaign and four of its neighboring counties is 15.26. 

sample_mean[ sample_mean['multi_unit'] >= 15.26 ].shape[0] / 5000

0.2212

**d)** Which is more unusual of the multi-unit rate of Champaign county or the average multi-unit rate of Champaign county and its four neighboring counties.  Explain, including numerical support in your justification.

In [18]:
county[(county['multi_unit'] >= 35.2) & (county['state'] == 'Illinois') ].shape[0] / county[county['state'] == 'Illinois'].shape[0]

0.0196078431372549

In [19]:
# The mean multi-rate of Champaign and four of its neighboring counties is 15.26. 

sample_mean[ sample_mean['multi_unit'] >= 15.26 ].shape[0] / 5000

0.2212

Multi-unit rate of Champaign county is 35.2. (from part 1b)

The proportion that a randomly selected county in Illinois that have a multi-unit rate that is at least as high as Champaign county's rate is approximately 0.02, which means only 2% of counties in Illinois have higher value of multi-unit rate than that of Champaign county. 

The average multi-unit rate of Champaign county and its four neighboring counties is 15.26. (from part 2b)

The proportion of samples of size 5 that have a mean multi-unit rate at least as large as the mean multi-unit rate of Champaign and four of its neighboring counties (15.26) is approximately 0.22, which means 22% of samples have higher mean multi-unit rate than Champaign and neighboring counties. 

Therefore, the multi-unit rate of Champaign county is more unusual. 

<hr>

## <u>Case Study</u>: Valentine's Day Spending

## 4. How much do college students spend on Valentine's Day?

A resident advisor aims to answer this question, so he surveys all residents who live in his dorm at the fictitious College University.  He saves this information a data file.

He has a few friends who claim that they do not spend very much on Valentine's Day.  In fact, the average spending of these friends is only $6.

The resident advisor, always a skeptic, wants to assess the claim of his friends: do they actually spend a small amount on Valentine's Day compared to his residents?  He knows that the $6 average of his friends shouldn't be compared to the population distribution for all of his residents, so he simulates a sampling distribution for the average spending of his residents on Valentine's Day.  He saves the results of this simulation to another data file.

**a)** Read in the two data files: vday1.csv and vday2.csv.  Report the number of observations for each file, and preview the first few rows of each file.

In [24]:
vday1 = pd.read_csv('vday1.csv')
print("The number of observations for vday1 is : ", vday1.shape[0])
vday1.head()


The number of observations for vday1 is :  214


Unnamed: 0  spending
0           0      2.95
1           1      8.14
2           2      5.36
3           3      2.12
4           4      1.79

In [25]:
vday2 = pd.read_csv('vday2.csv')
print("The number of observations for vday2 is : ", vday2.shape[0])
vday2.head()


The number of observations for vday2 is :  214


Unnamed: 0  spending
0           0      4.17
1           1      8.05
2           2      0.00
3           3      0.00
4           4     16.15

**b)** Unfortunately, the resident advisor realizes that he did not use the best naming system for his two files.  He forgets which file contains the population data and which file contains the simulated sampling distribution.  Help the resident advisor out by determining which data file contains which distribution.  Explain to him how you know that you are correct.

In [26]:
print(vday1['spending'].std())
print(vday2['spending'].std())

2.454647348831522
5.050620494085822


The first data, 'vday1.csv' contains the data of samples and the second data. 'vday2.csv' contains the data of population.

The standard deviation of 'vday1' data is smaller than that of 'vday2'. 

The value of standard deviation of population must be square root (n) (n = sample size) times than that of sample. Therefore, vday1 is contains samples while vday2 contains population.

**c)** Calculate the parameter values for the mean and standard deviation.

In [27]:
# parameter value of the mean
print (vday2['spending'].mean())

5.0285514018691595


In [28]:
# parameter value of the standard deviation
print (vday2['spending'].std())

5.050620494085822


**d)** The resident advisor has once again been a little careless with his record keeping.  He forgot to record how many friends he surveyed in order to calculate the $6 average spent on Valentine's Day.  Based on the available information, help him identify how many friends he surveyed.

In [38]:
sample_size = ((vday2['spending'].std()) / (vday1['spending'].std()))**2
sample_size

4.233614001002806

According to the explanation written in part b, I calculated what the sample size would be by two standard deviations. 

Based on this result, he might have surveyed 4 people.

**e)** Finally, are his friends especially stingy in terms of Valentine's Day spending, relative to the residents in his dorm?  Explain.

In [41]:
# the mean value of spending of vday1
print(vday1['spending'].mean())

# the percentage of vday1 spending is at least $6
print(vday1[vday1['spending']>= 6].shape[0] / vday1.shape[0])

5.015233644859812
0.32710280373831774


The percentage of sample means that have at least $6 as their mean value of spending is approximately 32.7%. 

Therefore, his friends are not stingy in terms of Valentine's Day spending because the amount ($6) is higher than the average spending of sample mean, and it is within 32.7% of the spending. 

<hr>

## <u>Case Study</u>: Super Bowl

## 5. Random Variables

Suppose that the number of touchdowns scored by a team in the Super Bowl follows a Poisson random variable with parameter $\mu = 2.4$ ($\mu$ is also sometimes called $\lambda$).

Information about Poisson random variables is contained within the scipy.stats package.  The code to import this is included below.

In [42]:
from scipy.stats import poisson

**a)** One important aspect of working with a programming language is the ability to get information about the packages.

To start, find the documentation for this package.  You may consider performing a web search where you include the package name (scipy.stats) and the specific distribution name.

Provide the url to the documentation that you find.

From reading through this documentation, is a Poisson random variable discrete or continuous?

A Poisson random variable is a discrete random variable.

**b)** Continuing to look through the documentation, what are the first and second methods listed in the table of available methods?

The first method is "rvs(mu, loc=0, size=1, random_state=None)" which stands for random variates. 

The second method is "pmf(k, mu, loc=0)" which stands for probability mass function. 

**c)** As of this homework's writing, you can bet on whether you think the Chiefs will score more than 2.5 touchdowns, with the bets favoring the Chiefs will score less than this number of touchdowns.

If the number of touchdowns for the Chiefs follows a Poisson distribution with parameter $\mu = 2.4$, what ist he probability that the Chiefs will score 2 touchdowns or less?

In [44]:
poisson.pmf(0,2.4,loc=0) + poisson.pmf(1, 2.4, loc = 0) + poisson.pmf(2, 2.4, loc = 0)

0.5697087466575105

**d)** Using the underlying Poisson distribution described above, randomly simulate the number of touchdowns scored for each team (two teams play).  You may use random states if you'd like, but you don't need to.

In [47]:
poisson.rvs(2.4, loc=0, size = 2, random_state = 123)

# The result from my random simulation is 2 and 5. 

array([2, 5], dtype=int64)

**e)** You can also bet on whether you think the total score of the game will be above or below 51 points.

We'll make a simplifying assumption that each touchdown will be worth 7 points.

Based on the simulation in part **d**, how many touchdowns would be scored in your simulated game?  How many points do those touchdowns correspond to?  Based only on the number of touchdowns scored, would the score of the game be above or below 51 points?

In [48]:
poisson.rvs(2.4, loc=0, size = 2, random_state = 123) * 7

array([14, 35], dtype=int64)

Based on the simulation in part d, the touchdowns would be 2 times and 5 times scored for each team. 

Those touchdowns will correspond to 14 points and 35 points each. 

Therefore, the score of the game would be below 51 points. 